content/posts/jspi.md

```diff
@@ -0,0 +1,494 @@
+
+### API compatibility with `run_sync`
+
+Calling a python function that uses `run_sync()` inside from JavaScript requires using `callPromising()` to enable stack switching.
+
+For instance, if you call the `do_something_with_request()` function from JavaScript:
+
+```js
+do_something_with_request = pyodide.globals.get("do_something_with_request");
+do_something_with_request("https://example.com");
+```
+
+It will raise:
+
+```py
+RuntimeError: Cannot stack switch because the Python entrypoint was a synchronous function. Use pyFunc.callPromising() to fix.
+```
+
+For API compatibility, you need to call it like
+
+```js
+await_async_http_request.callPromising("https://example.com")
+```
+
+This returns a promise, even though a synchronous Python function would
+ordinarily return the value directly.
+
+Also, calling the Python code that uses `run_sync()` requires using `pyodide.runPythonAsync()` instead of `pyodide.runPython()` to enable stack switching. 
+
+
+```js
+pyodide.runPython("run_sync(asyncio.sleep(1))"); // RuntimeError: Cannot stack switch ...
+pyodide.runPythonAsync("run_sync(asyncio.sleep(1))"); // Works fine
+```
+
+but as long as the Python entrypoint is an async function or invoked via
+`callPromising()` stack switching will be enabled. It is also possible to query
+whether or not stack switching is enabled with `pyodide.ffi.can_run_sync()`.
+
+## Using JSPI directly
+
+All of these examples are designed to work with NodeJS 24. Node must be started
+with `--experimental-wasm-stack-switching` since Node 24 was released from a
+version of v8 just before the feature gate was removed.
+
+We'll start with a basic example of the JSPI API. The full example is
+[here](https://github.com/hoodmane/jspi-blog-examples/tree/main/2-basic-example).
+
+Suppose we have an async JavaScript function:
+```js
+async function awaitAsyncHttpRequest(x) {
+    // Actually just sleep lol
+    console.log("JS: sleeping");
+    await sleep(1000);
+    console.log("JS: slept");
+    return x + 1;
+}
+```
+which we want to call from WebAssembly. From C's perspective, `awaitAsyncHttpRequest()`
+will return an `int`. We use it in the following C function:
+```C
+// This doesn't actually have anything to do with Python, but it would in our
+// application to the Pyodide code base.
+WASM_EXPORT("pythonFunction")
+void pythonFunction(int x) {
+    logString("About to call awaitAsyncHttpRequest");
+    int res = awaitAsyncHttpRequest(x);
+    logString("Got result:");
+    logInt(res);
+}
+```
+We compile and link this C file with clang. We compile with
+```sh
+clang -I../include -target wasm32-unknown-unknown -Wl,--no-entry -nostdlib -O2 \
+    basic.c -o basic.wasm
+```
+The build script is
+[here](https://github.com/hoodmane/jspi-blog-examples/blob/main/2-basic-example/build.sh).
+To instantiate the WebAssembly module, we need JavaScript definitions for the
+imports `sleep`, `logInt` and `logString`:
+```js
+function logInt(x) {
```

Comment:
While it is not an important part, some people might be confused about why using JS functions to print values in C. It is probably worth noting that we cannot use `stdlib` here (without Emscripten), so things like `printf` are not available.